File: YTP_2026_Hackthon_SakuraNaviBackend/app/infrastructure/db/models/resume.py
"""SQLAlchemy ORM model for the resumes table."""
import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSON, UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.infrastructure.db.base import Base


class ResumeModel(Base):
    __tablename__ = "resumes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(100), nullable=False)
    summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    skills: Mapped[Optional[list]] = mapped_column(JSON, nullable=True)  # [{"name": "...", "level": "..."}]
    work_experiences: Mapped[Optional[list]] = mapped_column(JSON, nullable=True)
    external_links: Mapped[Optional[list]] = mapped_column(JSON, nullable=True)  # [{"label": "...", "url": "..."}]
    expected_salary_min: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    expected_salary_max: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    expected_salary_currency: Mapped[Optional[str]] = mapped_column(String(3), nullable=True)
    work_time_range: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)  # {"start_time": "09:00", "end_time": "18:00", "work_time_type": "full_time"}
    is_primary: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )

File: YTP_2026_Hackthon_SakuraNaviApp/src/features/user-profile/types/resume.ts
export type WorkTimeType =
  | "full_time"
  | "part_time"
  | "internship"
  | "freelance";

export interface SkillInput {
  name: string;
  level?: string;
}

export interface WorkExperienceInput {
  company: string;
  position: string;
  description?: string;
  start_date: string;
  end_date?: string;
  is_current?: boolean;
}

export interface ExternalLinkInput {
  label: string;
  url: string;
}

export interface ExpectedSalaryInput {
  min: number;
  max: number;
  currency: string;
}

export interface WorkTimeRangeInput {
  start_time?: string;
  end_time?: string;
  work_time_type: WorkTimeType;
}

export interface Resume {
  id: string;
  user_id: string;
  title: string;
  summary: string | null;
  skills?: SkillInput[] | null;
  work_experiences?: WorkExperienceInput[] | null;
  external_links?: ExternalLinkInput[] | null;
  expected_salary?: ExpectedSalaryInput | null;
  work_time_range?: WorkTimeRangeInput | null;
  is_primary: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateResumePayload {
  title: string;
  summary?: string | null;
  skills?: SkillInput[] | null;
  work_experiences?: WorkExperienceInput[] | null;
  external_links?: ExternalLinkInput[] | null;
  expected_salary?: ExpectedSalaryInput | null;
  work_time_range?: WorkTimeRangeInput | null;
}

export interface UpdateResumePayload {
  title?: string | null;
  summary?: string | null;
  skills?: SkillInput[] | null;
  work_experiences?: WorkExperienceInput[] | null;
  external_links?: ExternalLinkInput[] | null;
  expected_salary?: ExpectedSalaryInput | null;
  work_time_range?: WorkTimeRangeInput | null;
}
